core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala

```diff
@@ -0,0 +1,123 @@
+    // +-----------------------------+----------------------------------------------+
+    // |    executor.launchTask()    |                                              |
+    // |                             | TaskRunner.run() begins                      |
+    // |                             |          ...                                 |
+    // |                             | execBackend.statusUpdate  // 1st time, #L240 |
+    // | executor.killAllTasks(true) |                                              |
+    // |                             |          ...                                 |
+    // |                             |  task = ser.deserialize   // #L253           |
+    // |                             |          ...                                 |
+    // |                             | execBackend.statusUpdate  // 2nd time, #L365 |
+    // |                             |          ...                                 |
+    // |                             |   TaskRunner.run() ends                      |
+    // |       check results         |                                              |
+    // +-----------------------------+----------------------------------------------+
+
+    val mockExecutorBackend = mock(classOf[ExecutorBackend])
+    when(mockExecutorBackend.statusUpdate(any(), any(), any()))
+      .thenAnswer(new Answer[Unit] {
+        var firstTime = true
+        override def answer(invocationOnMock: InvocationOnMock): Unit = {
+          if (firstTime) {
+            TestHelper.latch1.countDown()
+            // here between latch1 and latch2, executor.killAllTasks() is called
+            TestHelper.latch2.await()
+            firstTime = false
+          }
+          else {
+            val taskState = invocationOnMock.getArguments()(1).asInstanceOf[TaskState]
+            // save the returned `taskState` and `testFailedReason` into TestHelper
+            TestHelper.taskState = taskState
+            val taskEndReason = invocationOnMock.getArguments()(2).asInstanceOf[ByteBuffer]
+            TestHelper.testFailedReason = serializer.newInstance().deserialize(taskEndReason)
+            // let the main test thread to check `taskState` and `testFailedReason`
+            TestHelper.latch3.countDown()
+          }
+        }
+      })
+
+    val executor = new Executor("id", "localhost", mockEnv, userClassPath = Nil, isLocal = true)
+    // the task will be launched in a dedicated worker thread
+    executor.launchTask(mockExecutorBackend, 0, 0, "", serializedTask)
+
+    TestHelper.latch1.await()
+    executor.killAllTasks(true)
+    TestHelper.latch2.countDown()
+    TestHelper.latch3.await()
+
+    // `exception` should be `TaskKilled`; `taskState` should be `KILLED`
+    assert(TestHelper.testFailedReason === TaskKilled)
+    assert(TestHelper.taskState === TaskState.KILLED)
+  }
+}
+
+// Helps to test("SPARK-15963")
+private object TestHelper {
```

Comment:
can you rename this to `ExecutorSuiteHelper`?

also, would you mind making this a class instead of an object?  This is a pretty minor point -- the reason is that to flush out flaky tests, I sometimes like to run the tests in a loop:

```scala
(0 until 1000).foreach { idx => 
  test(s"... $idx){ ... }
}
```

and that doesn't work this as an object.